drivers/serial/uart_nrfx_uarte.c

```diff
@@ -829,36 +877,64 @@ static int uarte_nrfx_rx_enable(const struct device *dev, uint8_t *buf,
 	 * for the RXTO event after a call to uart_rx_disable() to discard
 	 * data from the UARTE internal RX FIFO.
 	 */
-	if (data->async->rx.enabled || data->async->rx.discard_fifo) {
+	if (rdata->enabled || rdata->discard_fifo) {
 		return -EBUSY;
 	}
 
-	data->async->rx.timeout = timeout;
-	data->async->rx.timeout_slab = timeout / RX_TIMEOUT_DIV;
+#ifdef CONFIG_UART_NRFX_UARTE_ENHANCED_RX
+#ifdef UARTE_HAS_FRAME_TIMEOUT
+	if (timeout && (timeout != SYS_FOREVER_US)) {
+		uint32_t baudrate = COND_CODE_1(CONFIG_UART_USE_RUNTIME_CONFIGURE,
+			(data->uart_config.baudrate), (cfg->baudrate));
 
-	data->async->rx.buf = buf;
-	data->async->rx.buf_len = len;
-	data->async->rx.offset = 0;
-	data->async->rx.next_buf = NULL;
-	data->async->rx.next_buf_len = 0;
+		rdata->timeout = K_USEC(timeout);
+		nrf_uarte_frame_timeout_set(uarte, us_to_bauds(baudrate, timeout));
+		nrf_uarte_shorts_enable(uarte, NRF_UARTE_SHORT_FRAME_TIMEOUT_STOPRX);
+	} else {
+		rdata->timeout = K_NO_WAIT;
+	}
+#else
+	rdata->timeout = (timeout && timeout == SYS_FOREVER_US) ?
+		K_NO_WAIT : K_USEC(timeout / RX_TIMEOUT_DIV);
+	rdata->idle_cnt = RX_TIMEOUT_DIV - 1;
```

Comment:
1. Why `- 1`? Timeout is split into `RX_TIMEOUT_DIV` parts and the `STOPRX` task is triggered when this counter is decreased to 0, so one such part earlier.
2. Wouldn't it be more convenient to initialize this counter to 0 (initialization is done in several places) and check if it reaches certain value before triggering the `STOPRX` task (this happens in one place only).
3. I got a bit confused. The commit message states that there will be no more `UART_RX_RDY` events with non-zero offsets. But because of this timeout split, it seems to me that they still can occur. Only when the frame timeout feature is used, they cannot, because the timeout is handled differently then. What am I missing?

Reply:
Changed to start with 0.

Reply:
It's still `RX_TIMEOUT_DIV - 1`, which I believe is incorrect, as the timeout is divided into RX_TIMEOUT_DIV parts.

And the commit message still contains:
> Note that new RX mode behaves a bit different because timeout always
triggers switch of buffers which means that there will be no
UART_RX_RDY events with non-zero offset. It also means that every
UART_RX_RDY will be followed by UART_RX_BUF_RELEASED.

which I think is true only when the hardware frame timeout is used.